### ISSUE TITLE
[Question] Concept Attribution Maps in CRAFT. How is it done in Xplique?

### ISSUE THREAD

**Issue by skumar-ml** [deel-ai/xplique#162]
I am interested to see how the implicit differentiation mentioned in CRAFT is implemented to create the Concept Attribution Map. Specifically, I want to know where the computation described in the paper (implicit diff w/ Jaxopt and the backprop) to find the jacobian d**U** / d**X** is done in Xplique, but I have been unable to find it. 

[This line](https://github.com/deel-ai/xplique/blob/master/xplique/concepts/craft.py#L716) gets the heatmaps, but tracing the call shows that we are retrieving **U** from the scikit-learn's NMF ([see here](https://github.com/deel-ai/xplique/blob/master/xplique/concepts/craft.py#L287)). Instead of doing implicit differentiation, it seems as if we're weighting embeddings (taken from **U**) according to the concept bank (**W**), and  we use the weight to encode the embedding's concept attribution score. Then, **U** is scaled up back to the original image resolution [in this line](https://github.com/deel-ai/xplique/blob/master/xplique/concepts/craft.py#L725) (as a way to find the pixel-level attribution, but I think this assumes that the embedding is spatially localized in the original image, which may hold for early/intermediate layers of CNNs) . Unless I've missed something, it seems to me that Xplique's implementation of CRAFT is not doing the implicit differentiation proposed in the original paper.

I've been able to find a TensorFlow implementation of the implicit differentiation in a separate repository [here](https://github.com/deel-ai/Craft/tree/main/craft/optim), but I wanted to see what the PyTorch implementation looked like. 

Your clarity on this would be appreciated! 

(After I understand how the Jacobian is obtained, I am also interested to know if the simple gradient is used as the pixel-level attribution, or if a more sophisticated gradient-based attribution method like Integrated Gradients is used. I could not find anything on this in the original CRAFT paper).

**Comment by lucashervier** [comment]
Hello there,

Sorry for the delay. I have been assigning the issue to @Agustin-Picard and @fredericboisnard who should be able to give you specific details. There are indeed some subtle differences to allow a more robust implementation in the Xplique library but I will let the ones concerned give you the specifics.

**Comment by Agustin-Picard** [comment]
Hi,

Sorry for the delay. I've been a bit busy lately. 

So, what you're saying is exactly right. We no longer do the implicit differentiation implementation as it can be quite compute-expensive. You already found our repo of the original implementation where we do use implicit differentiation (and require the `jax` and `jaxopt` libraries, as well as an interface to pass the implicit gradients computed in JAX to TF, which can be a bit cumbersome). We've decided to just include a simplified version of Craft (that we know should work about just as well) for maintainability and simplicity reasons. Hence, as you can imagine, we did not implement the "full" version of Craft (with `jaxopt`) in PyTorch. If you need to implement it yourself, I'd keep the implementation of how implicit differentiation is done in `jax` as it could get a bit tricky to code without all the helper toolboxes.

Regarding your last question, in the paper, once we have computed the gradients, and because we were mostly working on the last convolutional layer of a ResNet50, we applied GradCAM. Of course, we could have used SmoothGrad, Integrated Gradients, or any other gradient-based attribution method (or perturbation-based, even!), but doing so would have required more backward passes using implicit differentiation (to get to the **U** matrix), making it quite expensive to compute. Your intuition is right, though: we could have used any other attribution method in the toolbox!

I don't know if I have successfully addressed all of your questions. If not, I can provide further clarification.
